### ISSUE TITLE
Class member type inference inconsistency

### ISSUE THREAD

**Issue by amotzop** [microsoft/pyright#3353]
**Describe the bug**
Sometime it seems like the type of a class member can be precisely inferred, and is correctly inferred inside the classes `__init__`, but is mis-inferred outside of it.

**To Reproduce**
In the code below, the two calls to `reveal_type` return different results (it shows `Literal['']` inside `__init__` but `int | str` outside).

**Expected behavior**
I think even outside the `__init__` function the type should be inferred to be the same as in the end of the `__init__` function.

**Screenshots or Code**
```python
class Class:
    def __init__(self):
        self.member = 0
        self.member = ""
        reveal_type(self.member)

reveal_type(Class().member)
```

**VS Code extension or command-line**
This happens both in the VS Code extension and the command-line tool (version 1.1.237)

**Additional context**
The example above seems a bit strange, but it's simply the shortest one I found, this also happens in larger examples (where there really are more options for `self.member`'s type during `__init__`, but in the end they all collapse to a single type).

**Comment by erictraut** [comment]
The current behavior is both correct and consistent. The `reveal_type` reveals the locally-narrowed type of an expression. The instance variable `member` has no type declaration, so its type is inferred from all values assigned to it within the class. It is assigned twice in your example — an `int` and ` str` — so its inferred type is `int | str`. Locally within the `__init__` method, the type of the expression `self.member` is narrowed based on assignment, so its type changes throughout the code flow. If you insert another `reveal_type(self.member)` immediately after the `self.member = 0` assignment, you'll see that the revealed type is `Literal[0]` at that location.

If you want the instance variable `member` to have a different type, you can declare it by adding a type annotation.

```python
class Class:
    def __init__(self):
        self.member: str = ""
```

**Comment by amotzop** [comment]
That's what I thought, but what I meant was that you could infer the type of `self.member` from all it's possible values at the ends of all the methods of the class, instead of all assignments.
I think this kind of thing mostly happens when one assigns some temporary value to a member, only to update it later, e.g., doing some arithmetic on `int`s and then transforming the result to a string  and in such cases this type of inference seems to me to be more "correct" as the member is of type `str` and shouldn't be thought of as an `int` outside the relevant class method.

**Comment by erictraut** [comment]
It's not clear what the "ends of all the methods" means. Does "end" include only return statements? Yield statements? What if the method raises an exception or calls another method that raises an exception or uses an operator that implicitly calls a method that raises an exception? For type safety, a type checker needs to assume that all assignments to an instance variable contribute to its inferred type.
